File: Docs/05-simulator/샘플/epson_scan_api/scanner_wia.py
"""
scanner_wia.py  -  Epson scanner control via WIA (pywin32).

WIA works well with the EPSON ES-C320W (detected as a WIA Scanner).
Scanning is manual/on-demand: call scan() when paper is on the glass/ADF.
If there's no paper, the driver raises an error -> we surface it cleanly.

Needs: pip install pywin32   (run on the Windows PC)
"""

import logging
import os
import time
from pathlib import Path

# WIA constants
WIA_FORMAT_BMP = "{B96B3CAB-0728-11D3-9D7B-0000F81EF32E}"
WIA_FORMAT_PNG = "{B96B3CAF-0728-11D3-9D7B-0000F81EF32E}"
WIA_FORMAT_JPEG = "{B96B3CAE-0728-11D3-9D7B-0000F81EF32E}"

# WIA item property IDs
WIA_IPS_CUR_INTENT = 6146   # 1=color 2=grayscale 4=text(bw)
WIA_IPS_XRES = 6147
WIA_IPS_YRES = 6148
WIA_IPA_DATATYPE = 4103     # 0=bw threshold, 2=grayscale, 3=color
WIA_DPS_DOCUMENT_HANDLING_SELECT = 3088  # 1=FEEDER, 2=FLATBED

INTENT = {"color": 1, "gray": 2, "grayscale": 2, "bw": 4, "text": 4}
DATATYPE = {"color": 3, "gray": 2, "grayscale": 2, "bw": 0, "text": 0}


import contextlib


@contextlib.contextmanager
def _com():
    """Per-thread COM init (FastAPI runs requests on a threadpool)."""
    import pythoncom
    try:
        pythoncom.CoInitialize()
        inited = True
    except Exception:
        inited = False
    try:
        yield
    finally:
        if inited:
            try:
                pythoncom.CoUninitialize()
            except Exception:
                pass


def _dispatch(prog):
    import win32com.client
    return win32com.client.Dispatch(prog)


def list_scanners():
    """Return [{id, name}] of connected WIA scanners (Type==1)."""
    with _com():
        mgr = _dispatch("WIA.DeviceManager")
        out = []
        for i in range(1, mgr.DeviceInfos.Count + 1):
            info = mgr.DeviceInfos.Item(i)
            if info.Type != 1:  # 1 = Scanner
                continue
            name = ""
            try:
                for p in info.Properties:
                    if p.Name == "Name":
                        name = p.Value
            except Exception:
                pass
            out.append({"id": info.DeviceID, "name": name or info.DeviceID})
        return out


def _find_prop(props, pid):
    """This driver indexes Properties.Item() by POSITION, not by PropertyID,
    so locate the property by iterating and matching PropertyID."""
    try:
        n = props.Count
    except Exception:
        return None
    for i in range(1, n + 1):
        try:
            pr = props.Item(i)
            if int(pr.PropertyID) == int(pid):
                return pr
        except Exception:
            continue
    return None


def _set_prop(props, pid, value):
    pr = _find_prop(props, pid)
    if pr is None:
        return False
    try:
        pr.Value = value
        return True
    except Exception:
        return False


def _get_prop(props, pid, default=None):
    pr = _find_prop(props, pid)
    if pr is None:
        return default
    try:
        return pr.Value
    except Exception:
        return default


# device-level document-handling property IDs
WIA_DPS_DOCUMENT_HANDLING_CAPABILITIES = 3086
WIA_DPS_DOCUMENT_HANDLING_STATUS = 3087
CAP_FEEDER = 1
CAP_FLATBED = 2
STATUS_FEED_READY = 1
STATUS_FLAT_READY = 2


def scan(out_path, device_id=None, dpi=300, mode="gray", source="flatbed", fmt="bmp"):
    """
    Acquire one page to out_path. Returns out_path.
    Raises ScannerError with a clear message (e.g. no paper / no device).
    mode: color|gray|bw   source: flatbed|feeder   fmt: bmp|png|jpeg
    """
    import pythoncom
    with _com():
        mgr = _dispatch("WIA.DeviceManager")
        target = None
        for i in range(1, mgr.DeviceInfos.Count + 1):
            info = mgr.DeviceInfos.Item(i)
            if info.Type != 1:
                continue
            if device_id is None or info.DeviceID == device_id:
                target = info
                break
        if target is None:
            raise ScannerError("연결된 스캐너를 찾을 수 없습니다 (device_id=%s)." % device_id)

        device = target.Connect()

        caps = _get_prop(device.Properties, WIA_DPS_DOCUMENT_HANDLING_CAPABILITIES, 0) or 0
        has_feeder = bool(caps & CAP_FEEDER)
        has_flatbed = bool(caps & CAP_FLATBED)

        # choose source. This device may be ADF-only -> force feeder.
        want_feeder = str(source).lower() in ("feeder", "adf") or (has_feeder and not has_flatbed)
        sel = CAP_FEEDER if want_feeder else CAP_FLATBED
        _set_prop(device.Properties, WIA_DPS_DOCUMENT_HANDLING_SELECT, sel)

        # if using the feeder, make sure paper is actually detected
        if want_feeder:
            status = _get_prop(device.Properties, WIA_DPS_DOCUMENT_HANDLING_STATUS, 0) or 0
            if not (status & STATUS_FEED_READY):
                raise ScannerError("급지구에 용지가 감지되지 않습니다. 용지를 가이드에 맞춰 "
                                   "롤러가 물 때까지 끝까지 밀어 넣고 다시 시도하세요. (status=%s)" % status)

        item = device.Items.Item(1)
        xr = _set_prop(item.Properties, WIA_IPS_XRES, int(dpi))
        _set_prop(item.Properties, WIA_IPS_YRES, int(dpi))
        dt = _set_prop(item.Properties, WIA_IPA_DATATYPE, DATATYPE.get(mode, 2))
        try:
            _set_prop(item.Properties, WIA_IPS_CUR_INTENT, INTENT.get(mode, 2))
        except Exception:
            pass

        sel_now = _get_prop(device.Properties, WIA_DPS_DOCUMENT_HANDLING_SELECT)
        st_now = _get_prop(device.Properties, WIA_DPS_DOCUMENT_HANDLING_STATUS)
        logging.info("WIA scan: caps=%s select=%s status=%s xres_set=%s dtype_set=%s mode=%s dpi=%s",
                     caps, sel_now, st_now, xr, dt, mode, dpi)

        fmt_guid = {"bmp": WIA_FORMAT_BMP, "png": WIA_FORMAT_PNG, "jpeg": WIA_FORMAT_JPEG,
                    "jpg": WIA_FORMAT_JPEG}.get(str(fmt).lower(), WIA_FORMAT_BMP)

        image = None
        last_err = None
        for attempt in range(2):
            try:
                image = item.Transfer(fmt_guid)
                break
            except pythoncom.com_error as e:
                last_err = e
                hr = _extract_hresult(e)
                logging.warning("Transfer attempt %d failed: scode=0x%08X",
                                attempt + 1, (hr or 0) & 0xFFFFFFFF)
                if (hr or 0) & 0xFFFFFFFF == 0x80210003 and attempt == 0:
                    time.sleep(1.2)   # let the feeder settle, then retry once
                    continue
                raise ScannerError(_wia_error_text(hr) +
                                   (" [scode=0x%08X]" % (hr & 0xFFFFFFFF) if hr else "")) from e
        if image is None:
            hr = _extract_hresult(last_err) if last_err else None
            raise ScannerError(_wia_error_text(hr) +
                               (" [scode=0x%08X]" % (hr & 0xFFFFFFFF) if hr else ""))

        out_path = str(out_path)
        if Path(out_path).exists():
            Path(out_path).unlink()
        image.SaveFile(out_path)
        return out_path


def _extract_hresult(e):
    """Pull the HRESULT/scode out of a pywin32 com_error in a robust way."""
    try:
        if getattr(e, "excepinfo", None) and e.excepinfo[5]:
            return int(e.excepinfo[5])
    except Exception:
        pass
    for attr in ("hresult", "strerror"):
        v = getattr(e, attr, None)
        if isinstance(v, int):
            return v
    try:
        if e.args and isinstance(e.args[0], int):
            return e.args[0]
    except Exception:
        pass
    return None


def _wia_error_text(hr):
    table = {
        0x80210003: "스캐너에 용지가 없습니다 (ADF empty). 종이를 올리고 다시 시도하세요.",
        0x80210006: "스캐너가 준비되지 않았습니다 (busy/warming up).",
        0x80210001: "일반 스캔 오류입니다.",
        0x80210015: "스캐너를 사용할 수 없습니다 (offline/disconnected).",
        0x8021000A: "스캔 통신 오류입니다.",
        0x80210067: "스캔이 취소되었습니다.",
    }
    if hr is None:
        return "스캔 실패 (알 수 없는 오류)."
    return table.get(hr & 0xFFFFFFFF, "스캔 실패 (HRESULT 0x%08X)." % (hr & 0xFFFFFFFF))


class ScannerError(Exception):
    pass

File: Docs/05-simulator/샘플/Back Version/epson_scan_api_cs_20260626_/Services/ScannerWia.cs
using System.Runtime.InteropServices;
using System.Runtime.Versioning;

namespace EpsonScanApi.Services;

public class ScannerError(string msg) : Exception(msg);

[SupportedOSPlatform("windows")]
public static class ScannerWia
{
    // WIA format GUIDs
    private const string FmtBmp  = "{B96B3CAB-0728-11D3-9D7B-0000F81EF32E}";
    private const string FmtPng  = "{B96B3CAF-0728-11D3-9D7B-0000F81EF32E}";
    private const string FmtJpeg = "{B96B3CAE-0728-11D3-9D7B-0000F81EF32E}";

    // WIA item property IDs
    private const int WIA_IPS_CUR_INTENT  = 6146;
    private const int WIA_IPS_XRES        = 6147;
    private const int WIA_IPS_YRES        = 6148;
    private const int WIA_IPA_DATATYPE    = 4103;
    private const int WIA_DPS_DOCUMENT_HANDLING_SELECT       = 3088;
    private const int WIA_DPS_DOCUMENT_HANDLING_CAPABILITIES = 3086;
    private const int WIA_DPS_DOCUMENT_HANDLING_STATUS       = 3087;
    private const int CAP_FEEDER  = 1;
    private const int CAP_FLATBED = 2;
    private const int STATUS_FEED_READY = 1;

    private static readonly Dictionary<string, int> Intent   = new() { ["color"]=1,["gray"]=2,["grayscale"]=2,["bw"]=4,["text"]=4 };
    private static readonly Dictionary<string, int> Datatype = new() { ["color"]=3,["gray"]=2,["grayscale"]=2,["bw"]=0,["text"]=0 };
    private static readonly Dictionary<string, int> WiaHr = new()
    {
        [unchecked((int)0x80210003).ToString()] = 0,
        [unchecked((int)0x80210006).ToString()] = 1,
        [unchecked((int)0x80210001).ToString()] = 2,
        [unchecked((int)0x80210015).ToString()] = 3,
        [unchecked((int)0x8021000A).ToString()] = 4,
        [unchecked((int)0x80210067).ToString()] = 5,
    };
    private static readonly string[] WiaMessages =
    [
        "스캐너에 용지가 없습니다 (ADF empty). 종이를 올리고 다시 시도하세요.",
        "스캐너가 준비되지 않았습니다 (busy/warming up).",
        "일반 스캔 오류입니다.",
        "스캐너를 사용할 수 없습니다 (offline/disconnected).",
        "스캔 통신 오류입니다.",
        "스캔이 취소되었습니다.",
    ];

    public static List<Dictionary<string, string>> ListScanners()
    {
        List<Dictionary<string, string>>? result = null;
        Exception? error = null;
        var thread = new Thread(() =>
        {
            try
            {
                dynamic mgr = CreateWia();
                result = new List<Dictionary<string, string>>();
                int count = (int)mgr.DeviceInfos.Count;
                for (int i = 1; i <= count; i++)
                {
                    dynamic info = mgr.DeviceInfos.Item(i);
                    if ((int)info.Type != 1) continue;
                    string name = "";
                    try
                    {
                        foreach (dynamic p in info.Properties)
                            if ((string)p.Name == "Name") { name = (string)p.Value; break; }
                    }
                    catch { }
                    result.Add(new() { ["id"] = (string)info.DeviceID, ["name"] = name != "" ? name : (string)info.DeviceID });
                }
            }
            catch (Exception ex) { error = ex; }
        });
        thread.SetApartmentState(ApartmentState.STA);
        thread.Start();
        thread.Join();
        if (error != null)
            System.Runtime.ExceptionServices.ExceptionDispatchInfo.Capture(error).Throw();
        return result ?? [];
    }

    /// <summary>
    /// WIA COM은 STA(Single-Thread Apartment)에서만 안정적으로 동작.
    /// ASP.NET Core 스레드풀(MTA)에서 직접 호출하면 이미지 데이터가 비정상적으로 반환될 수 있음.
    /// Python의 pythoncom.CoInitialize() 동작을 재현.
    /// </summary>
    public static string Scan(string outPath, string? deviceId, int dpi, string mode, string source, string fmt)
    {
        string? result = null;
        Exception? error = null;
        var thread = new Thread(() =>
        {
            try { result = ScanCore(outPath, deviceId, dpi, mode, source, fmt); }
            catch (Exception ex) { error = ex; }
        });
        thread.SetApartmentState(ApartmentState.STA);
        thread.Start();
        thread.Join();
        if (error != null)
            System.Runtime.ExceptionServices.ExceptionDispatchInfo.Capture(error).Throw();
        return result!;
    }

    private static string ScanCore(string outPath, string? deviceId, int dpi, string mode, string source, string fmt)
    {
        dynamic mgr = CreateWia();
        dynamic? target = null;
        int count = (int)mgr.DeviceInfos.Count;
        for (int i = 1; i <= count; i++)
        {
            dynamic info = mgr.DeviceInfos.Item(i);
            if ((int)info.Type != 1) continue;
            if (deviceId == null || (string)info.DeviceID == deviceId) { target = info; break; }
        }
        if (target == null)
            throw new ScannerError($"연결된 스캐너를 찾을 수 없습니다 (device_id={deviceId}).");

        dynamic device = target.Connect();
        int caps = GetProp(device.Properties, WIA_DPS_DOCUMENT_HANDLING_CAPABILITIES, 0);
        bool hasFeeder  = (caps & CAP_FEEDER) != 0;
        bool hasFlatbed = (caps & CAP_FLATBED) != 0;

        bool wantFeeder = source.ToLower() is "feeder" or "adf" || (hasFeeder && !hasFlatbed);
        SetProp(device.Properties, WIA_DPS_DOCUMENT_HANDLING_SELECT, wantFeeder ? CAP_FEEDER : CAP_FLATBED);

        if (wantFeeder)
        {
            int status = GetProp(device.Properties, WIA_DPS_DOCUMENT_HANDLING_STATUS, 0);
            if ((status & STATUS_FEED_READY) == 0)
                throw new ScannerError("급지구에 용지가 감지되지 않습니다. 용지를 가이드에 맞춰 끝까지 밀어 넣고 다시 시도하세요.");
        }

        dynamic item = device.Items.Item(1);
        SetProp(item.Properties, WIA_IPS_XRES, dpi);
        SetProp(item.Properties, WIA_IPS_YRES, dpi);
        SetProp(item.Properties, WIA_IPA_DATATYPE, Datatype.GetValueOrDefault(mode, 2));
        try { SetProp(item.Properties, WIA_IPS_CUR_INTENT, Intent.GetValueOrDefault(mode, 2)); } catch { }

        string fmtGuid = fmt.ToLower() switch
        {
            "png"  => FmtPng,
            "jpeg" => FmtJpeg,
            "jpg"  => FmtJpeg,
            _      => FmtBmp,
        };

        if (File.Exists(outPath)) File.Delete(outPath);

        for (int attempt = 0; attempt < 2; attempt++)
        {
            try
            {
                dynamic image = item.Transfer(fmtGuid);
                image.SaveFile(outPath);
                return outPath;
            }
            catch (COMException ex) when (attempt == 0 && ((uint)ex.HResult == 0x80210003))
            {
                Thread.Sleep(1200);
            }
            catch (COMException ex)
            {
                throw new ScannerError(WiaErrorText(ex.HResult));
            }
        }
        throw new ScannerError("스캔 실패 (재시도 후에도 용지 없음).");
    }

    private static string WiaErrorText(int hr)
    {
        uint uhr = (uint)hr;
        return uhr switch
        {
            0x80210003 => WiaMessages[0],
            0x80210006 => WiaMessages[1],
            0x80210001 => WiaMessages[2],
            0x80210015 => WiaMessages[3],
            0x8021000A => WiaMessages[4],
            0x80210067 => WiaMessages[5],
            _ => $"스캔 실패 (HRESULT 0x{uhr:X8}).",
        };
    }

    private static dynamic CreateWia()
    {
        var t = Type.GetTypeFromProgID("WIA.DeviceManager", true)!;
        return Activator.CreateInstance(t)!;
    }

    private static void SetProp(dynamic props, int pid, int value)
    {
        try
        {
            int n = (int)props.Count;
            for (int i = 1; i <= n; i++)
            {
                dynamic p = props.Item(i);
                if ((int)p.PropertyID == pid) { p.Value = value; return; }
            }
        }
        catch { }
    }

    private static int GetProp(dynamic props, int pid, int def)
    {
        try
        {
            int n = (int)props.Count;
            for (int i = 1; i <= n; i++)
            {
                dynamic p = props.Item(i);
                if ((int)p.PropertyID == pid) return (int)p.Value;
            }
        }
        catch { }
        return def;
    }
}
